cloud/test/integration/crds/validation_test.go

```diff
@@ -0,0 +1,342 @@
+/*
+Copyright 2019 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crds
+
+import (
+	"encoding/json"
+	"os"
+	"testing"
+
+	"github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/apis/devices/v1alpha1"
+	deviceutils "github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/utils"
+	"github.com/kubeedge/kubeedge/cloud/test/integration/fixtures"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func buildCrdClient(t *testing.T) *rest.RESTClient {
+	kubeConfigPath := os.Getenv("KUBE_CONFIG")
+	kubeAPIServerURL := os.Getenv("KUBE_APISERVER_URL")
+	kubeConfig, err := clientcmd.BuildConfigFromFlags(kubeAPIServerURL, kubeConfigPath)
+	if err != nil {
+		t.Fatalf("Failed to build device CRD client, error is %v", err)
+		return nil
+	}
+	crdClient, err := deviceutils.NewCRDClient(kubeConfig)
+	if err != nil {
+		t.Fatalf("Failed to build device CRD client, error is %v", err)
+		return nil
+	}
+	return crdClient
+}
+
+func TestValidDeviceModel(t *testing.T) {
+	tests := map[string]struct {
+		name          string
+		deviceModelFn func() *v1alpha1.DeviceModel
+	}{
+		"valid bluetooth device model": {
+			deviceModelFn: func() *v1alpha1.DeviceModel {
+				deviceModel := fixtures.NewDeviceModelBluetooth("bluetooth-device-model", fixtures.DefaultNamespace)
+				return deviceModel
+			},
+		},
+		"valid modbus rtu device model": {
+			deviceModelFn: func() *v1alpha1.DeviceModel {
+				deviceModel := fixtures.NewDeviceModelModbus("modbus-device-model", fixtures.DefaultNamespace)
+				return deviceModel
+			},
+		},
+		"valid opc ua device model": {
+			deviceModelFn: func() *v1alpha1.DeviceModel {
+				deviceModel := fixtures.NewDeviceModelOpcUA("opcua-device-model", fixtures.DefaultNamespace)
+				return deviceModel
+			},
+		},
+	}
+
+	crdClient := buildCrdClient(t)
```

Comment:
I guess you want to ensure the crdClient is not `nil`, or else L82 will see some exception when the `buildCrdClient` return `nil`.

Reply:
in the buildCrdClient() method , if the client construction fails , t.Fatalf will generate an exception and the test won't be executed further. So line 82 won't be hit.

```sh
--- FAIL: TestValidDeviceModel (0.00s)
    validation_test.go:37: Failed to build device CRD client, error is invalid configuration: no configuration has been provided
```

Reply:
yes, I was misleading by line 38 where it return `nil` on error.